### PR TITLE
LSIF: Fix empty pages due to bloom filter

### PR DESCRIPTION
**Pull request by efritz**
When looking for global references, we open (in chunks for pagination) the dumps which we know to import a given package. This does **not** mean that they include the symbol that we're searching for. In order to determine this, we do an application-level bloom filter check to see (with very high likelihood) whether or not the identifier we're looking for is one that the dump imported.

This works well, but has the effect of removing nearly all or all of the dumps from a page of results. This is very weird from a user perspective: they ask for a page of references at cursor _c_ and are given an empty result set and a new cursor _c+1_. 

This PR changes this flow so that the server instead re-queries more data from the database until an entire page's worth of dumps have passed the bloom filter test. This may add zero or more additional db queries in the backend for a single request. I don't think this is pathological (this change makes it better as we don't send multiple empty pages back to the client), and we can measure and determine if we need to select _more_ from the db in order to batch some of this together as a performance enhancement.

This is a prerequisite to the implementation of [RFC 58](https://docs.google.com/document/d/1Ccp89K_Yxq6yesBmUvQFhlczcP-PQxb7-HsmqNnqbLs).

### CODE REVIEW

File: lsif/src/xrepo.ts [hunk 1]
```diff
@@ -506,31 +506,118 @@ export class XrepoDatabase {
...
+        identifier: string
+        /** The maximum number of repository records to return. */
+        offset: number
+        /** The number of repository records to skip. */
+        limit: number
+        /** The total number of results that can be returned. */
+        count: number
+    }): Promise<{ references: ReferenceModel[]; count: number; newOffset: number }> {
+        const references: ReferenceModel[] = []
+        while (references.length < limit && offset < count) {
```
Comment (chrismwendt) [inline lsif/src/xrepo.ts]: Is there a possibility of an infinite loop when `getPage` returns 0 results but the limit hasn't been hit? I haven't found an exit condition.

Comment (efritz) [reply lsif/src/xrepo.ts]: `offset` increases on every iteration

Comment (chrismwendt) [reply lsif/src/xrepo.ts]: Are you referring to `offset += scanned`? Is it not possible for `scanned` to be 0?

Comment (efritz) [reply lsif/src/xrepo.ts]: Yes; getPages returns the next page of results when we know there are at least `count` results total. If it returns an empty page then count was wrong in the first place. I can make this explicit just as a point structure/documentation.

Comment (efritz) [reply lsif/src/xrepo.ts]: Updated to make it a bit clearer.

Comment (chrismwendt) [reply lsif/src/xrepo.ts]: Yeah, I'd like to see an obvious check for 0 just in case. Ah, I see it, thanks for adding.

**Pull request merged**